File: apps/mobile/lib/features/auth/auth_repository.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:sign_in_with_apple/sign_in_with_apple.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../core/supabase.dart';

/// OAuth deep-link target. Must match, character-for-character: this constant, the
/// iOS `CFBundleURLSchemes` entry, the Android intent-filter (scheme + host), and
/// the Supabase Auth redirect allow-list (dashboard). supabase_flutter captures
/// the deep-link return and runs the PKCE exchange itself — no custom deep-link
/// handling code is needed.
const kAuthRedirect = 'com.brickback://login-callback';

/// Thin wrapper over `userClient.auth`. Auth lives on the **user** project only;
/// the catalog client is always anon (never send the user JWT there).
///
/// The app is fully usable signed-out — sign-in only unlocks premium cloud sync.
/// Google OAuth is wired first (matches whatabrick); Apple + email-OTP code paths
/// are present but require provider config on the Supabase dashboard to complete.
class AuthRepository {
  /// Opens Google in the system browser; supabase_flutter finishes the exchange.
  Future<void> signInWithGoogle() {
    return userClient.auth.signInWithOAuth(
      OAuthProvider.google,
      redirectTo: kAuthRedirect,
      authScreenLaunchMode: LaunchMode.externalApplication,
      queryParams: const {'prompt': 'select_account'},
    );
  }

  /// Apple sign-in. On Apple platforms this is the **native** flow — an
  /// `ASAuthorizationController` (via `sign_in_with_apple`) yields an identity token bound to a
  /// SHA-256 nonce, exchanged with Supabase via `signInWithIdToken` — the production-correct path
  /// App Store review requires when social login ships (oracle AuthRepository.swift:62-66,
  /// SignInView.swift:36-49,133-172). Elsewhere it falls back to the web OAuth flow.
  ///
  /// The end-to-end round-trip still depends on the Apple provider being configured on the
  /// Supabase dashboard (deferred server config) and the "Sign in with Apple" capability being
  /// added to the iOS Runner target; the client logic here is complete.
  Future<void> signInWithApple() async {
    final isApplePlatform = !kIsWeb &&
        (defaultTargetPlatform == TargetPlatform.iOS ||
            defaultTargetPlatform == TargetPlatform.macOS);
    if (!isApplePlatform) return _signInWithAppleWeb();

    final rawNonce = _generateNonce();
    final hashedNonce = sha256.convert(utf8.encode(rawNonce)).toString();
    final credential = await SignInWithApple.getAppleIDCredential(
      scopes: const [
        AppleIDAuthorizationScopes.email,
        AppleIDAuthorizationScopes.fullName,
      ],
      nonce: hashedNonce,
    );
    final idToken = credential.identityToken;
    if (idToken == null) {
      throw const AuthException('Apple sign-in returned no identity token.');
    }
    await userClient.auth.signInWithIdToken(
      provider: OAuthProvider.apple,
      idToken: idToken,
      nonce: rawNonce,
    );
  }

  /// Web OAuth fallback for non-Apple platforms (Android / web).
  Future<void> _signInWithAppleWeb() {
    return userClient.auth.signInWithOAuth(
      OAuthProvider.apple,
      redirectTo: kAuthRedirect,
      authScreenLaunchMode: LaunchMode.externalApplication,
    );
  }

  /// A cryptographically-random nonce; its SHA-256 rides in the Apple authorization request and
  /// the raw value is handed to Supabase to bind the returned identity token.
  static String _generateNonce([int length = 32]) {
    const charset = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-._';
    final rng = Random.secure();
    return List.generate(length, (_) => charset[rng.nextInt(charset.length)]).join();
  }

  /// Email magic-link / OTP: sends a link that returns via the same deep link.
  Future<void> signInWithEmail(String email) {
    return userClient.auth.signInWithOtp(
      email: email.trim(),
      emailRedirectTo: kAuthRedirect,
    );
  }

  /// Guarantee *some* session so the authenticated `join_party` RPC works for a guest, WITHOUT
  /// forcing a real sign-in: joining a party needs neither premium nor an account. A no-op if a
  /// session already exists (real or guest); otherwise it mints a transparent **anonymous**
  /// session, stamping [displayName] into `raw_user_meta_data.name` so the server's
  /// `_party_display_name()` shows the chosen name in the roster instead of "Builder" (oracle
  /// AuthRepository.swift:50-56). Requires anonymous sign-ins enabled on the user project.
  Future<void> ensureGuestSession({String? displayName}) async {
    if (currentUser != null) return;
    final name = displayName?.trim();
    await userClient.auth.signInAnonymously(
      data: (name != null && name.isNotEmpty) ? {'name': name} : null,
    );
  }

  Future<void> signOut() => userClient.auth.signOut();

  Session? get currentSession => userClient.auth.currentSession;
  User? get currentUser => userClient.auth.currentUser;

  /// True only for a **real** (non-anonymous) account. A guest who joined a party holds a
  /// transparent anonymous session — they have a `currentUser` (so the party RPCs work) but count
  /// as signed-OUT for premium, cloud sync, and the Profile UI (oracle AuthRepository.swift:35-40).
  bool get isSignedIn => currentUser != null && !isAnonymous;

  /// True when the only session is a transparent guest (anonymous) one.
  bool get isAnonymous => currentUser?.isAnonymous ?? false;
}

final authRepositoryProvider = Provider<AuthRepository>((ref) => AuthRepository());

/// Emits on every auth change (sign-in, sign-out, token refresh). The router and
/// sync controller both listen to this.
final authStateProvider = StreamProvider<AuthState>(
  (ref) => userClient.auth.onAuthStateChange,
);

/// The **real-account** signed-in state, recomputed on every auth change. A transparent guest
/// (anonymous) session reads as signed-out here, exactly like [AuthRepository.isSignedIn].
final isSignedInProvider = Provider<bool>((ref) {
  ref.watch(authStateProvider);
  return ref.read(authRepositoryProvider).isSignedIn;
});

File: apps/ios/BrickBackKit/Sources/BrickBackKit/Auth/AuthRepository.swift
import Foundation
import Supabase

/// OAuth deep-link target. Must match, character-for-character: this constant, the iOS
/// `CFBundleURLSchemes` entry, and the Supabase Auth redirect allow-list. Reused from the
/// Flutter app so the existing Supabase redirect config stays valid.
public let kAuthRedirect = "com.brickback://login-callback"

/// Thin wrapper over `userClient.auth`. Auth lives on the **user** project only; the catalog
/// client is always anon (never send the user JWT there). The app is fully usable signed-out —
/// sign-in only unlocks premium cloud sync.
///
/// S5 wires the three sign-in paths behind their UI:
/// - **Apple** — native `ASAuthorizationController` (driven from the SwiftUI
///   `SignInWithAppleButton`) yields an identity token + the nonce, exchanged here via
///   `signInWithIdToken(.apple, …)` — the production-correct path App Store review requires.
/// - **Google** — the web OAuth flow (`signInWithOAuth(.google, …)`) presented via
///   `ASWebAuthenticationSession` inside supabase-swift; the `com.brickback://login-callback`
///   redirect returns the session. Matches the Flutter app (no GoogleSignIn SDK dependency).
/// - **Email OTP** — `signInWithOTP` mails a magic link that returns via the same deep link;
///   `handleOpenURL` runs supabase-swift's PKCE exchange.
///
/// The live OAuth/SMTP round-trip depends on Supabase dashboard provider config (deferred to
/// S8); the client logic here is complete and testable behind the debug force-premium path.
public final class AuthRepository: @unchecked Sendable {
    private let client: SupabaseClient

    public init(client: SupabaseClient) {
        self.client = client
    }

    /// A real (non-anonymous) account. A guest who joined a party holds a transparent *anonymous*
    /// session — they have a `currentUserId` (so the party RPCs work) but count as signed-OUT for
    /// premium, cloud sync, and the Profile UI. Only a real sign-in unlocks those.
    public var isSignedIn: Bool {
        guard let user = client.auth.currentUser else { return false }
        return !user.isAnonymous
    }
    /// True when the only session is a transparent guest (anonymous) one.
    public var isAnonymous: Bool { client.auth.currentUser?.isAnonymous ?? false }
    public var currentUserId: String? { client.auth.currentUser?.id.uuidString.lowercased() }
    public var currentUserEmail: String? { client.auth.currentUser?.email }

    /// Guarantee *some* session so an authenticated-only RPC (`join_party`) works for a guest,
    /// WITHOUT forcing a real sign-in: joining a party needs neither premium nor an account. A
    /// no-op if already signed in (real or guest); otherwise it mints a transparent anonymous
    /// session, stamping `displayName` into `raw_user_meta_data.name` so the server's
    /// `_party_display_name()` shows the chosen name in the roster instead of "Builder". Requires
    /// anonymous sign-ins to be enabled on the user project.
    public func ensureGuestSession(displayName: String? = nil) async throws {
        if client.auth.currentUser == nil {
            let name = displayName?.trimmingCharacters(in: .whitespacesAndNewlines)
            let data: [String: AnyJSON]? = (name?.isEmpty == false) ? ["name": .string(name!)] : nil
            try await client.auth.signInAnonymously(data: data)
        }
    }

    // MARK: - Sign-in

    /// Complete a native Sign in with Apple: exchange the Apple-issued identity token (and the
    /// raw nonce that was SHA-256'd into the authorization request) for a Supabase session.
    public func signInWithApple(idToken: String, nonce: String) async throws {
        try await client.auth.signInWithIdToken(
            credentials: .init(provider: .apple, idToken: idToken, nonce: nonce)
        )
    }

    /// Open Google in an `ASWebAuthenticationSession`; supabase-swift finishes the PKCE exchange
    /// on the `com.brickback://login-callback` return and returns the session.
    public func signInWithGoogle() async throws {
        try await client.auth.signInWithOAuth(
            provider: .google,
            redirectTo: URL(string: kAuthRedirect),
            queryParams: [("prompt", "select_account")]
        )
    }

    /// Email magic-link / OTP: mails a link that returns via the same deep link.
    public func signInWithEmail(_ email: String) async throws {
        try await client.auth.signInWithOTP(
            email: email.trimmingCharacters(in: .whitespacesAndNewlines),
            redirectTo: URL(string: kAuthRedirect)
        )
    }

    /// Forward an incoming OAuth/OTP deep-link URL to supabase-swift, which runs the PKCE
    /// exchange and emits on `signInStates()`. Registered from the app root's `.onOpenURL`.
    public func handleOpenURL(_ url: URL) {
        client.auth.handle(url)
    }

    public func signOut() async throws {
        try await client.auth.signOut()
    }

    /// Emits the **real-account** signed-in state on every auth change (sign-in, sign-out, token
    /// refresh). The sync controller and Profile listen to this; a transparent guest (anonymous)
    /// session reads as signed-out here, exactly like `isSignedIn`. Yields the current state first.
    public func signInStates() -> AsyncStream<Bool> {
        AsyncStream { continuation in
            let task = Task { [client] in
                for await change in client.auth.authStateChanges {
                    continuation.yield(change.session?.user.isAnonymous == false)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}
